core/nightshade/src/nightshade_task.rs

```diff
@@ -0,0 +1,412 @@
+            // in a queue instead, and process it after we have the payload.
+
+            let gossip = Gossip::new(
+                self.owner_id,
+                author,
+                GossipBody::PayloadRequest(vec!(author)),
+                &self.owner_secret_key,
+            );
+            self.send_gossip(gossip);
+        }
+    }
+
+    fn process_gossip(&mut self, gossip: Gossip<P>) {
+        if !gossip.verify(&self.public_keys[gossip.sender_id]) {
+            return;
+        }
+
+        match gossip.body {
+            GossipBody::NightshadeStateUpdate(message) => self.process_message(*message),
+            GossipBody::PayloadRequest(authorities) => self.send_payloads(gossip.sender_id, authorities),
+            GossipBody::PayloadReply(payloads) => self.receive_payloads(gossip.sender_id, payloads),
+        }
+    }
+
+    fn send_payloads(&self, receiver_id: AuthorityId, authorities: Vec<AuthorityId>) {
+        let mut payloads = Vec::new();
+        for a in authorities {
+            if let Some(ref p) = self.authority_blocks[a] {
+                payloads.push(p.clone());
+            }
+        }
+        let gossip = Gossip::new(
+            self.owner_id,
+            receiver_id,
+            GossipBody::PayloadReply(payloads),
+            &self.owner_secret_key,
+        );
+        self.send_gossip(gossip);
+    }
+
+    fn receive_payloads(&mut self, sender_id: AuthorityId, payloads: Vec<SignedBlock<P>>) {
+        for signed_payload in payloads {
+            let authority_id = signed_payload.block.author();
+
+            // If the signed block is not properly signed by its author, we mark the sender as adversary.
+            if !signed_payload.verify(&self.public_keys[authority_id]) {
+                self.nightshade_as_mut_ref().set_adversary(sender_id);
+                continue;
+            }
+
+            if let Some(ref p) = self.authority_blocks[authority_id] {
+                if p.block.hash() != signed_payload.block.hash() {
+                    self.nightshade_as_mut_ref().set_adversary(authority_id);
+                    self.authority_blocks[authority_id] = None;
+                }
+            } else {
+                self.authority_blocks[authority_id] = Some(signed_payload);
+                self.missing_payloads -= 1;
+            }
+        }
+    }
+
+    /// Sends gossip to random authority peers.
+    fn gossip_state(&self) {
+        let my_state = self.state();
+
+        for i in 0..self.num_authorities {
+            if i != self.owner_id {
+                let message = Message {
+                    sender_id: self.owner_id,
+                    receiver_id: i,
+                    state: my_state.clone(),
+                };
+                self.send_state(message);
+            }
+        }
+    }
+
+    /// We need to have the payload for anything we want to endorse
+    /// TODO do it in a smarter way
+    fn collect_missing_payloads(&self) {
```

Comment:
So we collect the missing payloads by spamming the hell out of all peers? :)
Seems very inefficient and brutal.

Reply:
@mikhailOK is working on a better strategy to solve the issue of sharing payloads without overloading the network. For now, we agree on having this dummy strategy that works fine. Notice that this strategy is not good because payloads have a big intersection regarding transactions, and we want to avoid sending redundant information. But if we assume each payload is independent and the network is reliable (message will always arrive) this is the way to go, participant A sends its payload to participant B only once.